lib/rubygem_fs.rb

```diff
@@ -96,14 +111,25 @@ def remove(*keys)
     private
 
     def path_for(key, base = @base_dir)
-      key = key.sub(%r{^/+}, "")
+      key = key.to_s.sub(%r{^/+}, "")
       path = base.join(key).expand_path
-      raise UnsafePathError, "Unsafe path: #{key} is outside of base: #{@base_dir}" unless descendant?(path)
+      raise InvalidPathError, "key #{key.inspect} is outside of base #{base.inspect}" unless descendant?(path, base)
       path
     end
 
-    def descendant?(path)
-      path.to_s.start_with?(@base_dir.to_s) && path != @base_dir
+    def dir_for(prefix, base = @base_dir)
+      return base if prefix.blank?
+      raise InvalidPathError, "prefix #{prefix.inspect} must end in / to avoid ambiguous behavior" unless prefix.to_s.end_with?("/")
```

Comment:
When using a delimiter in s3, you need the prefix to end in a slash. I'm just enforcing it here in local so it works consistently.